### PR TITLE
Add linters and make a separate ci-lint GitHub Actions workflow

### PR DESCRIPTION
**Pull request by tjanez**

### CODE REVIEW

**Review by abukosek — COMMENTED** [review]
What is the reason for splitting part of the Makefile into a separate `common.mk` file?  This repo isn't as complex as oasis-core, so this seems unnecessary to me and just hides stuff.
Also, while doing that, you replaced `printf` with `echo`, those are not the same, `printf` should be used.

**Comment by tjanez** [comment]
> What is the reason for splitting part of the Makefile into a separate `common.mk` file? This repo isn't as complex as oasis-core, so this seems unnecessary to me and just hides stuff.

Two reasons:

1. I plan to add more helper functions that will help with versioning, Change Log and release process in the future, similar to what we have in Oasis Core and Oasis Core Ledger.
2. In the future, I plan to extract the common parts of the Makefiles to a separate tooling repo to keep the repo tooling DRY and simpler to maintain.

**Comment by abukosek** [comment]
> > What is the reason for splitting part of the Makefile into a separate `common.mk` file? This repo isn't as complex as oasis-core, so this seems unnecessary to me and just hides stuff.
> 
> Two reasons:
> 
> 1. I plan to add more helper functions that will help with versioning, Change Log and release process in the future, similar to what we have in Oasis Core and Oasis Core Ledger.
> 2. In the future, I plan to extract the common parts of the Makefiles to a separate tooling repo to keep the repo tooling DRY and simpler to maintain.

Ah, ok, that makes sense :)

**Comment by tjanez** [comment]
>  Also, while doing that, you replaced `printf` with `echo`, those are not the same, `printf` should be used.

Oh, yes, I forgot to remove the explicit `\n` characters at the end of each message.

So, I was planning on using `echo` to unify this across the repos. Are you ok with that?

**Comment by abukosek** [comment]
> So, I was planning on using `echo` to unify this across the repos. Are you ok with that?

I'd prefer we use `printf`, because it's more portable when using the ANSI escape codes for colors.

I actually switched from `echo` to `printf` because a user reported that the colors don't work on macOS (its `echo` apparently doesn't support the escape codes), whereas `printf` works everywhere.
I wanted to make a PR to update this in oasis-core as well, but got sidetracked by other stuff.

**Comment by tjanez** [comment]
> 
> I'd prefer we use `printf`, because it's more portable when using the ANSI escape codes for colors.
> 
> I actually switched from `echo` to `printf` because a user reported that the colors don't work on macOS (its `echo` apparently doesn't support the escape codes), whereas `printf` works everywhere.
> I wanted to make a PR to update this in oasis-core as well, but got sidetracked by other stuff.

Thanks for this info.

I would still prefer to have this unified and use `echo` to make it easier to transition to the separate tooling repo approach. Once we've migrated to that, we can weigh in on printf vs. echo.

Regarding colors not working on macOS, I guess we could just ask people to update their Bash with `brew install bash` to make it work.

**Review by kostko — COMMENTED** [review]
Can we make sure that gofumpt is used for linting? Also see [Oasis Core golangcli-lint configuration](https://github.com/oasisprotocol/oasis-core/blob/master/go/.golangci.yml).

**Comment by tjanez** [comment]
> Can we make sure that gofumpt is used for linting? Also see [Oasis Core golangcli-lint configuration](https://github.com/oasisprotocol/oasis-core/blob/master/go/.golangci.yml).

Yes, of course.

**Comment by abukosek** [comment]
> I would still prefer to have this unified and use `echo` to make it easier to transition to the separate tooling repo approach. Once we've migrated to that, we can weigh in on printf vs. echo.

There's really not much to weigh, `printf` works, `echo` does not :)

> Regarding colors not working on macOS, I guess we could just ask people to update their Bash with `brew install bash` to make it work.

That would not fix it, since the problem is with the OS `echo` command and not the Bash built-in.

I don't have strong feelings about this, since I don't use macOS, but I think it's bad to introduce a regression for this, especially since it was already fixed based on user input.

**Comment by tjanez** [comment]
> > I would still prefer to have this unified and use `echo` to make it easier to transition to the separate tooling repo approach. Once we've migrated to that, we can weigh in on printf vs. echo.
> 
> There's really not much to weigh, `printf` works, `echo` does not :)

I don't have a strong preference, but I don't think `printf` is superior in all cases. E.g. with `printf`, one needs to always suffix messages with `\n`.

>
> > Regarding colors not working on macOS, I guess we could just ask people to update their Bash with `brew install bash` to make it work.
> 
> That would not fix it, since the problem is with the OS `echo` command and not the Bash built-in.

In that case, let's just switch to always using Bash's built-in `echo`. It should work for Linux users and macOS users. The latter will just need to update the pre-installed ancient version of Bash.

**Comment by tjanez** [comment]
https://github.com/oasisprotocol/oasis-core-rosetta-gateway/blob/fefb884060570525e033aa5ae60bf16ff8c7ab48/services/operations.go#L281-L285

@abukosek, @pro-wh, is this TODO comment still applicable?

If yes, we should file an issue so we don't forget about it.

**Comment by tjanez** [comment]
@kostko, this is ready for another review.

I did some further Go code refactoring that enables us using more golangci-lint's linters.

**Comment by abukosek** [comment]
> > > I would still prefer to have this unified and use `echo` to make it easier to transition to the separate tooling repo approach. Once we've migrated to that, we can weigh in on printf vs. echo.
> > 
> > 
> > There's really not much to weigh, `printf` works, `echo` does not :)
> 
> I don't have a strong preference, but I don't think `printf` is superior in all cases. E.g. with `printf`, one needs to always suffix messages with `\n`.
> 
> > > Regarding colors not working on macOS, I guess we could just ask people to update their Bash with `brew install bash` to make it work.
> > 
> > 
> > That would not fix it, since the problem is with the OS `echo` command and not the Bash built-in.
> 
> In that case, let's just switch to always using Bash's built-in `echo`. It should work for Linux users and macOS users. The latter will just need to update the pre-installed ancient version of Bash.

You're missing the point here -- `printf` is portable and works on both platforms without any other changes.

Telling users to upgrade their system `bash` just because we're too lazy to put `\n` at the end of strings is very poor user experience :P
Also, upgrading the system `bash` on macOS can break system scripts, so we really shouldn't advise users to do that, as it might wreck their system entirely.

File: services/operations.go [hunk 1]
```diff
@@ -305,11 +363,170 @@ func readOasisCurrencyNeg(amount *types.Amount) (*quantity.Quantity, error) {
...
+
+// decodeOpsToTransaction decodes the Oasis transaction type from the given
+// Rosetta operations.
+func decodeOpsToTransaction(ops []*types.Operation) string {
+	switch {
+	case len(ops) == 1:
+		switch {
+		case ops[0].Type == OpBurn &&
+			ops[0].Account.SubAccount == nil:
+			return "staking_burn"
```
Comment (kostko) [inline services/operations.go:494]: All of these should be typed constants (e.g., return type should not be a generic string).

File: services/operations.go [hunk 2]
```diff
@@ -533,165 +646,205 @@ func (m *transactionToOperationMapper) EmitFeeOps() {
...
+	return nil
+}
+
+// EmitTxOps emits the required transaction-specific operations.
+func (m *transactionToOperationMapper) EmitTxOps() error {
+	switch m.tx.Method {
+	case staking.MethodTransfer:
+		if err := m.emitTransferOps(); err != nil {
+			return err
+		}
```
Comment (kostko) [inline services/operations.go:835]: Why not just `return m.emitTransferOps()` and similar below?

File: tests/check-prep/main.go [hunk 3]
```diff
@@ -136,18 +130,23 @@ func main() {
...
+			[]byte("balance/"+types.Hash(&testEntityAccountIdentifier)+"/"+types.Hash(services.OasisCurrency)),
+			storageEncode(
+				&struct {
+					Account *types.AccountIdentifier `json:"account"`
+					Amount  *types.Amount            `json:"amount"`
+					Block   *types.BlockIdentifier   `json:"block"`
+				}{
+					Account: &testEntityAccountIdentifier,
+					Amount: &types.Amount{
+						// https://github.com/oasisprotocol/oasis-core/blob/v20.8.2/go/oasis-node/cmd/genesis/genesis.go#L534
```
Comment (kostko) [inline tests/check-prep/main.go:143]: Why does this refer to some very old tag, maybe we should update the link?

File: services/operations.go (hunk 1 above)
Comment (tjanez) [reply services/operations.go]: Agreed. Done.

File: services/operations.go (hunk 2 above)
Comment (tjanez) [reply services/operations.go]: Indeed. Done.

File: tests/check-prep/main.go (hunk 3 above)
Comment (tjanez) [reply tests/check-prep/main.go]: I've changed this and exposed the test entity's general balance as the `TestEntityAmount` variable in `tests/common`.

**Review by kostko — APPROVED** [review]

**Comment by pro-wh** [comment]
> https://github.com/oasisprotocol/oasis-core-rosetta-gateway/blob/fefb884060570525e033aa5ae60bf16ff8c7ab48/services/operations.go#L281-L285
> 
> @abukosek, @pro-wh, is this TODO comment still applicable?
> 
> If yes, we should file an issue so we don't forget about it.

filed #105

File: tests/check-prep/main.go (hunk 3 above)
Comment (pro-wh) [reply tests/check-prep/main.go]: it's good practice to use permalinks so that the line numbers don't drift away.

**Pull request merged**